src/app/shared/services/doctors.service.ts

```diff
@@ -1,32 +1,160 @@
+  ngOnDestroy(): void {
+    this.subsription.unsubscribe();
+  }
+
+  loadData(): void {
+    this.subsription = this.db
+      .list('/doctors')
+      .valueChanges()
+      .subscribe((data) => {
+        this.firebaseDoctors = data as Doctor[];
+        this.updateVisibleDoctors();
+        this.calculateTopDoctors();
+      });
+
+    this.loadSpecialties();
   }
 
-  getSpecialties(): Observable<string[]> {
-    return this.doctors.pipe(
-      map((doctors) =>
-        Array.from(new Set(doctors.map((doctor) => doctor.specialization)))
-      )
+  loadSpecialties(): void {
+    this.db.database.ref('specialties').once('value', (snap) => {
+      this.firebaseSpecialties = snap.val();
+      const specialties = Object.entries(snap.val()).map(([id, name]) => ({
+        id,
+        name: name as string,
+        checked: false,
+      }));
+      this.visibleSpecialtiesSubject.next(specialties);
+    });
+  }
+
+  private updateVisibleDoctors(): void {
+    const specialties = this.visibleSpecialtiesSubject.value
+      .filter((v) => v.checked)
+      .map((v) => v.id);
+
+    const hasSearchFilter = !!this.searchQuerySubject.value;
+    const searchLowerStr = this.searchQuerySubject.value?.toLowerCase() || '';
+    const hasSpecialtiesFilter = !!specialties && specialties.length > 0;
+
+    const filteredDoctors = this.firebaseDoctors.filter((doc) => {
+      if (hasSearchFilter && !doc.name.toLowerCase().includes(searchLowerStr)) {
+        return false;
+      }
+      if (hasSpecialtiesFilter && !specialties?.includes(doc.specialization)) {
+        return false;
+      }
+      return true;
+    });
+
+    if (this.sortingTypeSubject.value) {
+      switch (this.sortingTypeSubject.value) {
+        case SortingTypes.name: {
+          filteredDoctors.sort((a, b) => a.name.localeCompare(b.name));
+          break;
+        }
+        case SortingTypes.experience: {
+          filteredDoctors.sort((a, b) => b.experience - a.experience);
+          break;
+        }
+      }
+    }
+
+    this.paginationSizeSubject.next(
+      Math.ceil(filteredDoctors.length / DOCTORS_PER_PAGE)
     );
+
+    const start = this.paginationCurrentSubject.value * DOCTORS_PER_PAGE;
+    const end = (this.paginationCurrentSubject.value + 1) * DOCTORS_PER_PAGE;
+
+    const cuttedDoctorsArray = filteredDoctors.slice(start, end);
+
+    const mappedSpecialties = this.formatData(cuttedDoctorsArray);
+
+    this.visibleDoctorsSubject.next(mappedSpecialties);
+  }
+
+  private calculateTopDoctors(): void {
+    const topDoctors = this.firebaseDoctors.filter((doctor) => doctor.top);
+    const formatedDoctors = this.formatData(topDoctors);
+    this.topDoctorsSubject.next(formatedDoctors);
+  }
+
+  private formatData(array: Doctor[]): Doctor[] {
+    return array.map((d) => ({
+      ...d,
+      specialization: get(
+        this.firebaseSpecialties,
+        d.specialization,
+        'unknown'
+      ),
+    }));
+  }
+
+  onPageChange(page: number): void {
```

Comment:
naming methods with `on` in service seems confusing, since it's not directly binded with view events. Try use verb + subject naming, like `changePage` and etc.